Add torch.__config__.show(), reporting detailed version of all libraries.

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#18579 Add torch.__config__.show(), reporting detailed version of all libraries.**

Fixes #111.

Example output:

```
PyTorch built with:
  - GCC 4.8
  - Intel(R) Math Kernel Library Version 2019.0.3 Product Build 20190125 for Intel(R) 64 architecture applications
  - Intel(R) MKL-DNN v0.18.1 (Git Hash 7de7e5d02bf687f971e7668963649728356e0c20)
  - OpenMP 201107 (a.k.a. OpenMP 3.1)
  - NNPACK is enabled
  - CUDA Runtime 9.2
  - NVCC architecture flags: -gencode;arch=compute_52,code=sm_52
  - CudNN 7.10.2  (built against CUDA 9.1)
  - Magma 2.5.0
  - Build settings: BLAS=MKL, BUILD_TYPE=Release, CXX_FLAGS=  -Wno-deprecated -fvisibility-inlines-hidden -fopenmp -O2 -fPIC -Wno-narrowing -Wall -Wextra -Wno-missing-field-initializers -Wno-type-limits -Wno-array-bounds -Wno-unknown-pragmas -Wno-sign-compare -Wno-unused-parameter -Wno-unused-variable -Wno-unused-function -Wno-unused-result -Wno-strict-overflow -Wno-strict-aliasing -Wno-error=deprecated-declarations -Wno-error=pedantic -Wno-error=redundant-decls -Wno-error=old-style-cast -Wno-unused-but-set-variable -Wno-maybe-uninitialized, DISABLE_NUMA=1, PERF_WITH_AVX=1, PERF_WITH_AVX2=1, USE_CUDA=True, USE_EXCEPTION_PTR=1, USE_GFLAGS=OFF, USE_GLOG=OFF, USE_MKL=ON, USE_MKLDNN=ON, USE_MPI=OFF, USE_NCCL=True, USE_NNPACK=True, USE_OPENMP=ON,
```

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

gh-metadata: pytorch pytorch 18579 gh/ezyang/48/head

Differential Revision: [D14778507](https://our.internmc.facebook.com/intern/diff/D14778507)

ezyang committed:
Add torch.backends.detailed_version(), reporting detailed version of all libraries.

Example output:

```
PyTorch built with:
  - GCC 4.8
  - Intel(R) Math Kernel Library Version 2019.0.3 Product Build 20190125 for Intel(R) 64 architecture applications
  - Intel(R) MKL-DNN v0.18.1 (Git Hash 7de7e5d02bf687f971e7668963649728356e0c20)
  - OpenMP 201107 (a.k.a. OpenMP 3.1)
  - CUDA Runtime 9.2.0
  - CuDNN 7.1.2
```

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

soumith commented:
this fixes https://github.com/pytorch/pytorch/issues/111

You might want to rename the methods slightly to match numpy, if you think it makes sense

ezyang commented:
Makes sense to me!

dzhulgakov commented on torch/csrc/Module.cpp:
```diff
@@ -299,6 +299,13 @@ PyObject *THPModule_hasDistributed(PyObject *_unused)
 #endif
 }
 
+static PyObject *THPModule_detailedVersion(PyObject *module)
```
nit: shall we start using pybind universally?

apaszke replied:
For things like this one sure, but I don't trust pybind enough to feel ok with using it for our performance-critical eager functions.

dzhulgakov reviewed (approved):
Would be nice to unify it with Caffe2's build options here: caffe2/core/macros.h.in caffe2/core/common.cc

Specifically, if you follow new format, I'd add the following compilation flags:
- BUILD_TYPE
- CXX_FLAGS
- BLAS (as we will likely replace the ugly THBlas soon)
- USE_NNPACK
- USE_FBGEMM (as we're adding quantization support now) - and include its version

ezyang commented:
> Would be nice to unify it with Caffe2's build options here: caffe2/core/macros.h.in caffe2/core/common.cc

Thanks, was not aware of this.

ezyang commented:
> USE_FBGEMM (as we're adding quantization support now) - and include its version

fbgemm does not appear to have any useful versions. Determined by looking at `Fbgemm.h` and searching for `configure_file` in the cmake

ezyang committed:
Update on "Add torch.__config__.show(), reporting detailed version of all libraries."

Add torch.__config__.show(), reporting detailed version of all libraries.

Fixes #111.

Example output:

```
PyTorch built with:
  - GCC 4.8
  - Intel(R) Math Kernel Library Version 2019.0.3 Product Build 20190125 for Intel(R) 64 architecture applications
  - Intel(R) MKL-DNN v0.18.1 (Git Hash 7de7e5d02bf687f971e7668963649728356e0c20)
  - OpenMP 201107 (a.k.a. OpenMP 3.1)
  - CUDA Runtime 9.2.0
  - CuDNN 7.1.2
```

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

gh-metadata: pytorch pytorch 18579 gh/ezyang/48/head

ezyang committed:
Windows fix and rebase on "Add torch.__config__.show(), reporting detailed version of all libraries."

Add torch.__config__.show(), reporting detailed version of all libraries.

Fixes #111.

Example output:

```
PyTorch built with:
  - GCC 4.8
  - Intel(R) Math Kernel Library Version 2019.0.3 Product Build 20190125 for Intel(R) 64 architecture applications
  - Intel(R) MKL-DNN v0.18.1 (Git Hash 7de7e5d02bf687f971e7668963649728356e0c20)
  - OpenMP 201107 (a.k.a. OpenMP 3.1)
  - CUDA Runtime 9.2.0
  - CuDNN 7.1.2
```

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

gh-metadata: pytorch pytorch 18579 gh/ezyang/48/head

ezyang committed:
Fix doc coverage test on "Add torch.__config__.show(), reporting detailed version of all libraries."

Add torch.__config__.show(), reporting detailed version of all libraries.

Fixes #111.

Example output:

```
PyTorch built with:
  - GCC 4.8
  - Intel(R) Math Kernel Library Version 2019.0.3 Product Build 20190125 for Intel(R) 64 architecture applications
  - Intel(R) MKL-DNN v0.18.1 (Git Hash 7de7e5d02bf687f971e7668963649728356e0c20)
  - OpenMP 201107 (a.k.a. OpenMP 3.1)
  - CUDA Runtime 9.2.0
  - CuDNN 7.1.2
```

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

gh-metadata: pytorch pytorch 18579 gh/ezyang/48/head

ezyang committed:
Fix HIP error on "Add torch.__config__.show(), reporting detailed version of all libraries."

Add torch.__config__.show(), reporting detailed version of all libraries.

Fixes #111.

Example output:

```
PyTorch built with:
  - GCC 4.8
  - Intel(R) Math Kernel Library Version 2019.0.3 Product Build 20190125 for Intel(R) 64 architecture applications
  - Intel(R) MKL-DNN v0.18.1 (Git Hash 7de7e5d02bf687f971e7668963649728356e0c20)
  - OpenMP 201107 (a.k.a. OpenMP 3.1)
  - CUDA Runtime 9.2.0
  - CuDNN 7.1.2
```

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

gh-metadata: pytorch pytorch 18579 gh/ezyang/48/head

ppwwyyxx commented:
For cuda/cudnn, etc, could build-time version be different from run-time version in its minor version number or patch number?
If yes, it would be great to have access to both build-time version and run-time version available for debugging (perhaps with `cudnnGetVersion` & `cudaRuntimeGetVersion`)

ezyang commented:
@ppwwyyxx Makes sense to me!

ezyang committed:
Update on "Add torch.__config__.show(), reporting detailed version of all libraries."

Add torch.__config__.show(), reporting detailed version of all libraries.

Fixes #111.

Example output:

```
PyTorch built with:
  - GCC 4.8
  - Intel(R) Math Kernel Library Version 2019.0.3 Product Build 20190125 for Intel(R) 64 architecture applications
  - Intel(R) MKL-DNN v0.18.1 (Git Hash 7de7e5d02bf687f971e7668963649728356e0c20)
  - OpenMP 201107 (a.k.a. OpenMP 3.1)
  - CUDA Runtime 9.2.0
  - CuDNN 7.1.2
```

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

gh-metadata: pytorch pytorch 18579 gh/ezyang/48/head

ezyang commented:
Now also with nvcc arch information

ezyang committed:
Update on "Add torch.__config__.show(), reporting detailed version of all libraries."

Add torch.__config__.show(), reporting detailed version of all libraries.

Fixes #111.

Example output:

```
PyTorch built with:
  - GCC 4.8
  - Intel(R) Math Kernel Library Version 2019.0.3 Product Build 20190125 for Intel(R) 64 architecture applications
  - Intel(R) MKL-DNN v0.18.1 (Git Hash 7de7e5d02bf687f971e7668963649728356e0c20)
  - OpenMP 201107 (a.k.a. OpenMP 3.1)
  - CUDA Runtime 9.2.0
  - CuDNN 7.1.2
```

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

gh-metadata: pytorch pytorch 18579 gh/ezyang/48/head

ezyang committed:
Add missing miopen header on "Add torch.__config__.show(), reporting detailed version of all libraries."

Add torch.__config__.show(), reporting detailed version of all libraries.

Fixes #111.

Example output:

```
PyTorch built with:
  - GCC 4.8
  - Intel(R) Math Kernel Library Version 2019.0.3 Product Build 20190125 for Intel(R) 64 architecture applications
  - Intel(R) MKL-DNN v0.18.1 (Git Hash 7de7e5d02bf687f971e7668963649728356e0c20)
  - OpenMP 201107 (a.k.a. OpenMP 3.1)
  - CUDA Runtime 9.2.0
  - CuDNN 7.1.2
```

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

gh-metadata: pytorch pytorch 18579 gh/ezyang/48/head

ezyang committed:
Fix syntax error in ROCm on "Add torch.__config__.show(), reporting detailed version of all libraries."

Add torch.__config__.show(), reporting detailed version of all libraries.

Fixes #111.

Example output:

```
PyTorch built with:
  - GCC 4.8
  - Intel(R) Math Kernel Library Version 2019.0.3 Product Build 20190125 for Intel(R) 64 architecture applications
  - Intel(R) MKL-DNN v0.18.1 (Git Hash 7de7e5d02bf687f971e7668963649728356e0c20)
  - OpenMP 201107 (a.k.a. OpenMP 3.1)
  - CUDA Runtime 9.2.0
  - CuDNN 7.1.2
```

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

gh-metadata: pytorch pytorch 18579 gh/ezyang/48/head

dzhulgakov reviewed (commented):
So, do you want to also include the following:
BUILD_TYPE
CXX_FLAGS
BLAS (as we will likely replace the ugly THBlas soon)
USE_NNPACK
USE_FBGEMM (as we're adding quantization support now) - and include its version

Even if FBGEMM doesn't have exported version, just the fact whether we build with it or not, same for NNPACK.

Even for MKL and others - I'd suggest to explicitly adding a line "MKL: none" to the output

ezyang commented:
@dzhulgakov I'm assuming part of your comment is an accidental copy paste from the earlier comment.

Right now, I print all of the existing Caffe2 settings as part of the `Build settings` key, filtering out keys which are set to an empty string. Do you want each of these to be given their own line in the output? Also, are you asking to stop filtering out empty string keys? (I feel they add unnecessary noise to the output)

ezyang committed:
Add missing miopen include on "Add torch.__config__.show(), reporting detailed version of all libraries."

Add torch.__config__.show(), reporting detailed version of all libraries.

Fixes #111.

Example output:

```
PyTorch built with:
  - GCC 4.8
  - Intel(R) Math Kernel Library Version 2019.0.3 Product Build 20190125 for Intel(R) 64 architecture applications
  - Intel(R) MKL-DNN v0.18.1 (Git Hash 7de7e5d02bf687f971e7668963649728356e0c20)
  - OpenMP 201107 (a.k.a. OpenMP 3.1)
  - CUDA Runtime 9.2.0
  - CuDNN 7.1.2
```

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

gh-metadata: pytorch pytorch 18579 gh/ezyang/48/head

dzhulgakov reviewed (approved):
Apologies, I was looking at the old code version (GH UI is confusing)

ezyang committed:
Shut up error: no case matching constant switch condition  on "Add torch.__config__.show(), reporting detailed version of all libraries."

Add torch.__config__.show(), reporting detailed version of all libraries.

Fixes #111.

Example output:

```
PyTorch built with:
  - GCC 4.8
  - Intel(R) Math Kernel Library Version 2019.0.3 Product Build 20190125 for Intel(R) 64 architecture applications
  - Intel(R) MKL-DNN v0.18.1 (Git Hash 7de7e5d02bf687f971e7668963649728356e0c20)
  - OpenMP 201107 (a.k.a. OpenMP 3.1)
  - CUDA Runtime 9.2.0
  - CuDNN 7.1.2
```

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

gh-metadata: pytorch pytorch 18579 gh/ezyang/48/head

vadimkantorov commented:
would this be useful for `collect_env` script?

ppwwyyxx commented:
@ezyang 
>   - CudNN 7.10.2  (built against CUDA 9.1)

There is no cudnn 7.10.2.

Cudnn version numbers work a bit different from cuda. 7102 means 7.1.2.

ezyang commented:
@ppwwyyxx Oops! Thanks for the catch.
